contracts/src/v0.8/dev/ocr2dr/OCR2DRRegistry.sol

```diff
@@ -0,0 +1,783 @@
+    // Note that we need the s_consumers map to be able to directly check if a
+    // consumer is valid without reading all the consumers from storage.
+    address[] consumers;
+  }
+  // Note a nonce of 0 indicates an the consumer is not assigned to that subscription.
+  mapping(address => mapping(uint64 => uint64)) /* consumer */ /* subId */ /* nonce */
+    private s_consumers;
+  mapping(uint64 => SubscriptionConfig) /* subId */ /* subscriptionConfig */
+    private s_subscriptionConfigs;
+  mapping(uint64 => Subscription) /* subId */ /* subscription */
+    private s_subscriptions;
+  // We make the sub count public so that its possible to
+  // get all the current subscriptions via getSubscription.
+  uint64 private s_currentSubId;
+  // s_totalBalance tracks the total link sent to/from
+  // this contract through onTokenTransfer, cancelSubscription and oracleWithdraw.
+  // A discrepancy with this contract's link balance indicates someone
+  // sent tokens using transfer and so we may need to use recoverFunds.
+  uint96 private s_totalBalance;
+  event SubscriptionCreated(uint64 indexed subId, address owner);
+  event SubscriptionFunded(uint64 indexed subId, uint256 oldBalance, uint256 newBalance);
+  event SubscriptionConsumerAdded(uint64 indexed subId, address consumer);
+  event SubscriptionConsumerRemoved(uint64 indexed subId, address consumer);
+  event SubscriptionCanceled(uint64 indexed subId, address to, uint256 amount);
+  event SubscriptionOwnerTransferRequested(uint64 indexed subId, address from, address to);
+  event SubscriptionOwnerTransferred(uint64 indexed subId, address from, address to);
+
+  error InvalidRequestConfirmations(uint32 have, uint32 min, uint32 max);
+  error GasLimitTooBig(uint32 have, uint32 want);
+  error NumWordsTooBig(uint32 have, uint32 want);
+  error DonAlreadyRegistered(address don);
+  error NoSuchDon(address don);
+  error InvalidLinkWeiPrice(int256 linkWei);
+  error InsufficientGasForConsumer(uint256 have, uint256 want);
+  error NoCorrespondingRequest();
+  error IncorrectCommitment();
+  error BlockhashNotInStore(uint256 blockNum);
+  error PaymentTooLarge();
+  error Reentrant();
+
+  mapping(address => bool) /* DON */ /* is DON allowed */
+    private s_allowedDons;
+  address[] private s_dons;
+  mapping(address => uint96) /* oracle */ /* LINK balance */
+    private s_withdrawableTokens;
+  struct Commitment {
+    OCR2DRRegistryInterface.RequestBilling billing;
+    address don;
+    uint96 donFee;
+    uint96 registryFee;
+  }
+  mapping(bytes32 => Commitment) /* requestID */ /* Commitment */
+    private s_requestCommitments;
+  event DonRegistered(address indexed don);
+  event DonDeregistered(address indexed don);
+  event BillingStart(
+    address indexed don,
+    bytes32 requestId,
+    uint64 indexed subscriptionId,
+    uint32 callbackGasLimit,
+    address indexed sender
+  );
+  event BillingEnd(bytes32 indexed requestId, uint96 payment, bool success);
```

Comment:
Can we add `subscriptionId` to `BillingEnd` event? This may be useful for future observability work to capture all billing events for a given subscription.